ironfish/src/blockchain/blockchain.ts

```diff
@@ -1346,6 +1349,151 @@ export class Blockchain {
     }
   }
 
+  private async updateAssetsFromBlock(block: Block, tx: IDatabaseTransaction): Promise<void> {
+    for (const transaction of block.transactions) {
+      await this.processMintsToAssetsStore(transaction, tx)
+      await this.processBurnsToAssetsStore(transaction, tx)
+    }
+  }
+
+  private async processMintsToAssetsStore(
+    transaction: Transaction,
+    tx: IDatabaseTransaction,
+  ): Promise<void> {
+    for (const { asset, value } of transaction.mints()) {
+      const assetIdentifier = asset.identifier()
+      const existingAsset = await this.assets.get(assetIdentifier, tx)
+
+      let createdTransactionHash = transaction.hash()
+      let supply = BigInt(0)
+      if (existingAsset) {
+        createdTransactionHash = existingAsset.createdTransactionHash
+        supply = existingAsset.supply
+      }
+
+      await this.assets.put(
+        assetIdentifier,
+        {
+          createdTransactionHash,
+          metadata: asset.metadata(),
+          name: asset.name(),
+          nonce: asset.nonce(),
+          owner: asset.owner(),
+          supply: supply + value,
+        },
+        tx,
+      )
+    }
+  }
+
+  private async processBurnsToAssetsStore(
+    transaction: Transaction,
+    tx: IDatabaseTransaction,
+  ): Promise<void> {
+    for (const { asset, value } of transaction.burns()) {
+      const assetIdentifier = asset.identifier()
+      const existingAsset = await this.assets.get(assetIdentifier, tx)
+      Assert.isNotUndefined(existingAsset, 'Cannot burn undefined asset from the database')
+
+      const existingSupply = existingAsset.supply
+      const supply = existingSupply - value
+      Assert.isTrue(supply >= BigInt(0), 'Invalid burn value')
+
+      await this.assets.put(
+        assetIdentifier,
+        {
+          createdTransactionHash: existingAsset.createdTransactionHash,
+          metadata: asset.metadata(),
+          name: asset.name(),
+          nonce: asset.nonce(),
+          owner: asset.owner(),
+          supply,
+        },
+        tx,
+      )
+    }
+  }
+
+  private async undoAssetChangesFromBlock(
```

Comment:
nomenclature thing but we use the term disconnect to indicate a disconnection operation from the main chain